p2p/peer.go

```diff
@@ -214,97 +214,280 @@ type PeerUpdate struct {
+//   (or erroring if already connected).
+// - Ready: routing is up, broadcasts a PeerStatusUp peer update to subscribers.
+// - Disconnected: peer disconnects, unmarking as connected.
+//
+// For an inbound connection, the flow is as follows:
+// - Accepted: successfully accepted connection, marking as connected (or erroring
+//   if already connected).
+// - Ready: routing is up, broadcasts a PeerStatusUp peer update to subscribers.
+// - Disconnected: peer disconnects, unmarking as connected.
+//
+// We track dialing and connected states independently. This allows us to accept
+// an inbound connection while the router is also dialing an outbound
+// connection, which will cause the dialer to eventually error (when attempting
+// to mark the peer as connected). This also avoids race conditions where
+// multiple goroutines may end up dialing a peer if an incoming connection was
+// briefly accepted and disconnected while we were also dialing.
+type peerManager struct {
+	mtx           sync.Mutex
+	store         *peerStore
+	dialing       map[NodeID]bool
+	connected     map[NodeID]bool
+	subscriptions map[*PeerUpdatesCh]*PeerUpdatesCh // keyed by struct identity (address)
+}
+
+// newPeerManager creates a new peer manager.
+func newPeerManager(store *peerStore) *peerManager {
+	return &peerManager{
+		store:         store,
+		dialing:       map[NodeID]bool{},
+		connected:     map[NodeID]bool{},
+		subscriptions: map[*PeerUpdatesCh]*PeerUpdatesCh{},
+	}
+}
+
+// Add adds a peer to the manager, given as an address. If the peer already
+// exists, the address is added to it.
+func (m *peerManager) Add(address PeerAddress) error {
+	if err := address.Validate(); err != nil {
+		return err
+	}
+	peerID := address.NodeID()
+
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	peer, err := m.store.Get(peerID)
+	if err != nil {
+		return err
+	}
+	if peer == nil {
+		peer = newPeerInfo(peerID)
+	}
+	if peer.AddAddress(address) {
+		return m.store.Set(peer)
+	}
+	return nil
 }
 
-// newPeerStore creates a new peer store.
-func newPeerStore() *peerStore {
-	return &peerStore{
-		peers:   map[NodeID]*peerInfo{},
-		claimed: map[NodeID]bool{},
+// Subscribe subscribes to peer updates. The caller must consume the peer
+// updates in a timely fashion and close the subscription when done, since
+// delivery is guaranteed and will block peer connection/disconnection
+// otherwise.
+func (m *peerManager) Subscribe() *PeerUpdatesCh {
+	// FIXME: We may want to use a size 1 buffer here. When the router
+	// broadcasts a peer update it has to loop over all of the
+	// subscriptions, and we want to avoid blocking and waiting for a
+	// context switch before continuing to the next subscription. This also
+	// prevents tail latencies from compounding across updates. We also want
+	// to make sure the subscribers are reasonably in sync, so it should be
+	// kept at 1. However, this should be benchmarked first.
+	peerUpdates := NewPeerUpdates(make(chan PeerUpdate))
```

Comment:
I think it's safe to use a buffer of size 1 here for now. Any reason you chose to stick with unbuffered for now? I guess the only downside is we cannot safely reason about ordering of peer updates.

Reply:
You requested that this should be unbuffered for now in https://github.com/tendermint/tendermint/pull/5831#discussion_r553978323, until we can benchmark this.

Reply:
You're right! My bad 😆